### PR TITLE
Custom vendors

### PR DESCRIPTION
**Pull request by Tlaltecuhtli**
:cl:
add: custom vendors
add: changed the vendomat stock
/:cl:

it replaces&expands the never used chef custom input thing with a new vendor which accepts all kind of stuff
adds a price tagger which is used to set prices on a n y t h i n g

vendor has to be broken to eject things for free to prevent people crowbaring it

ingame view:

### CODE REVIEW

File: code/modules/research/designs/autolathe_designs.dm [hunk 1]
```diff
@@ -959,3 +959,19 @@
 	build_path = /obj/item/stack/tile/circuit/red
 	category = list("initial", "Misc")
 	maxstack = 50
+
+/datum/design/price_tagger
+	name = "Price Tagger"
+	id = "price_tagger"
+	build_type = AUTOLATHE
+	materials = list(/datum/material/iron = 1500, /datum/material/glass = 500))
```
Comment (bobbahbrown) [inline code/modules/research/designs/autolathe_designs.dm]: accidental trailing right brace?

File: code/modules/research/designs/autolathe_designs.dm [hunk 2]
```diff
@@ -959,3 +959,19 @@
...
+	build_path = /obj/item/price_tagger
+	category = list("initial", "Misc")
+
+/datum/design/custom_vendor_refill
+	name = "Custom Vendor Refill"
+	id = "custom_vendor_refill"
+	build_type = AUTOLATHE
+	materials = list(/datum/material/iron = 5000, /datum/material/glass = 2000))
```
Comment (bobbahbrown) [inline code/modules/research/designs/autolathe_designs.dm]: accidental trailing right brace?

File: code/modules/vending/_vending.dm [hunk 3]
```diff
@@ -759,33 +796,51 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+	name = "Custom Vendor"
+	icon_state = "robotics"
+	icon_deny = "robotics-deny"
+	payment_department = NO_FREEBIES
+	refill_canister = /obj/item/vending_refill/custom
+	custom = TRUE
+
+/obj/machinery/vending/custom/Destroy()
+	for(var/obj/item/I in contents)
+		I.forceMove(get_turf(src))
```
Comment (nemvar) [inline code/modules/vending/_vending.dm]: forceMove doesn't accept nulls. It's also wasteful to do get_turf() for ever single item.
Yes, I know this is copypasta, doesn't mean you don't have to fix it.

File: code/modules/vending/_vending.dm [hunk 4]
```diff
@@ -341,7 +349,7 @@ GLOBAL_LIST_EMPTY(vending_products)
 		.[record.product_path] += record.amount
 
 /obj/machinery/vending/crowbar_act(mob/living/user, obj/item/I)
-	if(!component_parts)
+	if(!component_parts && custom)
```
Comment (nemvar) [inline code/modules/vending/_vending.dm]:
```suggestion
	if(!component_parts || custom)

File: code/modules/vending/_vending.dm [hunk 5]
```diff
@@ -644,6 +670,17 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+/// proc to see who is the owner of the custom vendor
+/obj/machinery/vending/proc/isowner(mob/user)
+	. = FALSE
+	if(custom)
+		var/mob/living/carbon/human/H
+		var/obj/item/card/id/C
+		if(ishuman(user))
+			H = user
+			C = H.get_idcard(TRUE)
+			if(C.registered_account && C.registered_account == private_a)
```
Comment (bobbahbrown) [inline code/modules/vending/_vending.dm]: null reference on failing to get ID card

File: code/modules/vending/_vending.dm [hunk 6]
```diff
@@ -369,6 +377,29 @@ GLOBAL_LIST_EMPTY(vending_products)
 	if(panel_open && is_wire_tool(I))
 		wires.interact(user)
 		return
+
+	if(custom)
```
Comment (bobbahbrown) [inline code/modules/vending/_vending.dm]: unless im reading this wrong you should add a last return so you don't check for the other refill canisters as you can just return always at the end of this logic (if there was a private account linked, but the user wasn't the owner)

File: code/modules/vending/_vending.dm [hunk 7]
```diff
@@ -552,22 +571,29 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+				say("You do not possess the funds to purchase this.")
+			else
+				account.adjust_money(-price)
+				var/datum/bank_account/D = private_a
+				if(D)
+					D.adjust_money(price)
+				use_power(5)
+
+				vending_machine_input[N] = max(vending_machine_input[N] - 1, 0)
+				for(var/obj/O in contents)
```
Comment (nemvar) [inline code/modules/vending/_vending.dm]: Don't do this. You already did this exact loop a few lines ago. Instead just cache the item and check if it's still in the vendor here.

File: code/modules/vending/_vending.dm [hunk 8]
```diff
@@ -484,38 +496,45 @@ GLOBAL_LIST_EMPTY(vending_products)
 			if(vending_machine_input[O] > 0)
 				var/N = vending_machine_input[O]
 				dat += "<a href='byond://?src=[REF(src)];dispense=[sanitize(O)]'>Dispense</A> "
-				dat += "<B>[capitalize(O)] ($[default_price]): [N]</B><br>"
+				if(isowner(user))
+					price = "FREE"
+				else
+					for(var/obj/T in contents)
```
Comment (Qustinnus) [inline code/modules/vending/_vending.dm]: what is this loop even

File: code/modules/vending/_vending.dm [hunk 9]
```diff
@@ -369,6 +377,29 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+			if(V.recorded)
+				slogan_list += "[V.recorded]"
+			return
+		if(!private_a)
+			var/mob/living/carbon/human/H
+			var/obj/item/card/id/C
+			if(ishuman(user))
+				H = user
+				C = H.get_idcard(TRUE)
+				if(C && C.registered_account)
```
Comment (Qustinnus) [inline code/modules/vending/_vending.dm]: C?. or u die

**Comment by wesoda25** [comment]
shits fucking gas

**Comment by PKPenguin321** [comment]
Always wanted this

**Comment by Michael-Ashfield** [comment]
Cant wait to sell OPs the nuke disk

File: code/modules/vending/_vending.dm (hunk 8 above)
Comment (Tlaltecuhtli) [reply code/modules/vending/_vending.dm]: move along citizen

**Comment by cacogen** [comment]
ui looks like assholes but i really like the idea

**Comment by Time-Green** [comment]
now I can finally start selling hulk injectors this is really good

**Comment by necromanceranne** [comment]
Finally.

Also the image gives a fantastic idea for an admeme thing.

Lootboxes here we come.

**Comment by Shadowflame909** [comment]
Please tell me the money goes into IDs. Bro as a money I'm gonna start selling survival pens in medbay for double the cost and tendril loot I don't want.

This'll also make admin trader means 10x better. Imagine an admin bluespacing in a custom vendor where every adminbus item is 10k.

Speedmerge

File: code/modules/vending/_vending.dm [hunk 10]
```diff
@@ -759,33 +796,54 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+	icon_deny = "robotics-deny"
+	max_integrity = 400
+	payment_department = NO_FREEBIES
+	refill_canister = /obj/item/vending_refill/custom
+	custom = TRUE
+
+/obj/machinery/vending/custom/Destroy()
+	var/turf/T = get_turf(src)
+	for(var/obj/item/I in contents)
+		I?.forceMove(T)
```
Comment (nemvar) [inline code/modules/vending/_vending.dm]: That's not what I meant. T can be null and I will never be null.

**Comment by optimumtact** [comment]
can you set negative prices on things to boost your own income?
that seems like an exploit

**Comment by PKPenguin321** [comment]
Yeah that is kind of odd. It would be cool if you could put bounties on items and have the machine pay out for them though, so you could automate both buying and selling. Like for example you could set the machine to accept any insulated gloves and give a credit reward for it.

**Comment by MCterra10** [comment]
If you make it so the vendor can purchase items, please implement some kind of credit escrow system so that you can't offer to buy X at 10k credits, withdraw that money from your ID and not pay them. That or your balance goes negative and you're in debt

File: code/modules/vending/_vending.dm [hunk 11]
```diff
@@ -369,6 +377,37 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+	if(custom)
+		if(!private_a)
+			var/mob/living/carbon/human/H
+			var/obj/item/card/id/C
+			if(ishuman(user))
+				H = user
+				C = H.get_idcard(TRUE)
+				if(C?.registered_account)
+					private_a = C.registered_account
+					say("The [src] has been linked to [C].")
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/vending/_vending.dm]: I always forget but I think src handles The.

At best this needs `\The`

**Comment by ExcessiveUseOfCobblestone** [comment]
You know all the clothing vendors use the loading system too right (so that you can put clothes you don't want back in)?

**Comment by ExcessiveUseOfCobblestone** [comment]
Also why did you make a `custom` var? You have a subtype, overwrite the procs...

File: code/modules/vending/_vending.dm [hunk 12]
```diff
@@ -387,42 +392,23 @@ GLOBAL_LIST_EMPTY(vending_products)
...
-				else
-					denied_items++
-			if(denied_items)
-				to_chat(user, "<span class='warning'>[src] refuses some items!</span>")
-			if(loaded)
-				to_chat(user, "<span class='notice'>You insert [loaded] dishes into [src]'s chef compartment.</span>")
-				updateUsrDialog()
+	if(canLoadItem(I) && !custom)
+		loadingAttempt(I,user)
+		updateUsrDialog() //can't put this on the proc above because we spam it below
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/vending/_vending.dm]: where did this code go

File: code/modules/vending/_vending.dm [hunk 13]
```diff
@@ -529,15 +514,13 @@ GLOBAL_LIST_EMPTY(vending_products)
...
-
 	usr.set_machine(src)
-
 	if((href_list["dispense"]) && (vend_ready))
 		var/N = href_list["dispense"]
 		if(vending_machine_input[N] <= 0) // Sanity check, there are probably ways to press the button when it shouldn't be possible.
 			return
 		vend_ready = 0
-		if(ishuman(usr) && onstation)
+		if(ishuman(usr))
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/vending/_vending.dm]: ??

File: code/modules/vending/_vending.dm [hunk 14]
```diff
@@ -759,33 +746,200 @@ GLOBAL_LIST_EMPTY(vending_products)
...
   */
-/obj/machinery/vending/proc/canLoadItem(obj/item/I,mob/user)
-	return FALSE
-/**
-  * Is the passed in user allowed to load this vending machines compartments
-  *
-  * Arguments:
-  * * user - mob that is doing the loading of the vending machine
-  */
-/obj/machinery/vending/proc/compartmentLoadAccessCheck(mob/user)
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/vending/_vending.dm]: Where did this code go

**Review by ExcessiveUseOfCobblestone — COMMENTED** [review]
what r u doing

File: code/modules/vending/_vending.dm (hunk 12 above)
Comment (Tlaltecuhtli) [reply code/modules/vending/_vending.dm]: chef thing has been replaced by custom vendor

File: code/modules/vending/_vending.dm (hunk 14 above)
Comment (Tlaltecuhtli) [reply code/modules/vending/_vending.dm]: chef thing no longer used

File: code/modules/vending/_vending.dm (hunk 12 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/vending/_vending.dm]: this is not exclusive to chef items.

File: code/modules/vending/_vending.dm (hunk 14 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/vending/_vending.dm]: honestly should have kept it and overwritten this proc for custom...

File: code/modules/vending/_vending.dm (hunk 12 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/vending/_vending.dm]: actually they are in practice but the ability to bar certain people to loading from a vendor that isn't a personal one seems like a valid usecase.

**Comment by Tlaltecuhtli** [comment]
@ExcessiveUseOfCobblestone the stuff us back and unused do you need anything else?

**Comment by ExcessiveUseOfCobblestone** [comment]
more weekends to play games and actively avoid looking at the code

File: code/modules/vending/_vending.dm [hunk 15]
```diff
@@ -759,33 +790,202 @@ GLOBAL_LIST_EMPTY(vending_products)
...
+	payment_department = NO_FREEBIES
+	refill_canister = /obj/item/vending_refill/custom
+	/// where the money is sent
+	var/datum/bank_account/private_a
+	/// max number of items that the custom vendor can hold
+	var/max_loaded_items = 20
+
+/// proc to see who is the owner of the custom vendor
+
+/obj/machinery/vending/custom/proc/isowner(mob/user)
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/vending/_vending.dm]: can't you use `compartmentLoadAccessCheck` here instead?

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]

**Comment by FantasticFwoosh** [comment]
I have been waiting all month for this uuuurrrrrgghhhh

**Comment by PKPenguin321** [comment]
You fool, you're supposed to leave all the files checked when fixing conflicts

**Comment by ExcessiveUseOfCobblestone** [comment]
how do u mess up this badly

**Comment by PKPenguin321** [comment]
it's a noob trap when fixing conflicts
the last step where you commit will have every single file that was updated listed as a changed file, youre supposed to include all of these files so that the merge conflict will automatically fix
however, youre also led to believe that you should only include files that you changed, so people unaware of the merge conflict fixing thing will often uncheck most of the files. it totally breaks everything

**Comment by ExcessiveUseOfCobblestone** [comment]
github :b:esktop is the noob trap

**Pull request closed without merging**